thepracticaldev/dev.to
# Add path_redirects table and internal view

## What type of PR is this? (check all applicable)

## Description
**PR Summary**
This PR creates the backend for `path_redirects` (migration, model, and updated seed file) and creates a new area on internal for admins to access it appropriately.

I included a `version` column so we can keep an eye on many times we're updating a re-direct. Constantly updating a URL with a new `301` destination hurts SEO, so we want to be mindful of that.

The `source` column is so we know the origin of the `new_path` field. It can be one of two things:
1. `"service"` - the `new_path` was determined programmatically.
2. `"admin"` - the `new_path` was set by an admin in the internal panel.

**Steps to test**
1. Re-seed your DB or create `PathRedirect` records manually.
2. [Make sure your user has admin access.](https://docs.dev.to/internal/#internal-guide)
3. Fire-up the app and visit [http://localhost:3000/internal/path_redirects](http://localhost:3000/internal/path_redirects).
4. View, update, create, and destroy `PathRedirects`. Don't forget to try and break it (i.e. submit blank data in the form to make sure it throws an error 😎 .

**Admin usage**
Admins can:
- Create `PathRedirects`.
- Edit only the `new_path` of an existing `PathRedirect`.
- Destroy `PathRedirects`.
- View `PathRedirects`.

The version number is automatically incremented anytime the `new_path` field is updated. The `source` field is automatically set to `"admin"` for any action made on a record via the internal/admin panel. 

And lastly, the version number is color-coded:
- The color is green when the version is <= 3.
- The color is yellow when the version is <= 5.
- The color is red when the version is > 5.

**Project Summary**
_I'll be copy and pasting this summary on each PR related to this project_

We are working on adding better handling for 404 pages to improve SEO. To start, we're focusing on Articles. When an Article is unpublished, we'll add logic to store a new place to redirect to. Then, if a user tries to visit an Article that is going to result in a `404` response, we'll check if we have a defined redirect and do that accordingly.
- [x] [Create a redirects table to hold redirects and expose it to admins](https://github.com/thepracticaldev/dev.to/issues/8079)
- [ ] [Create an `ArticleRedirectService`](https://github.com/thepracticaldev/dev.to/issues/8080)
- [ ] [Create a Sidekiq worker to call `ArticleRedirectService` when an Article is unpublished](https://github.com/thepracticaldev/dev.to/issues/8081)
- [ ] [Create a controller concern and handle Article redirects](https://github.com/thepracticaldev/dev.to/issues/8083)
- [ ] [Log redirects to Datadog](https://github.com/thepracticaldev/dev.to/issues/8082)
- [ ] [Create a JS redirect notice](https://github.com/thepracticaldev/dev.to/issues/8084)

## Related Tickets & Documents
https://github.com/thepracticaldev/dev.to/issues/8079

## Mobile & Desktop Screenshots/Recordings
**The index/show view**
- An admin can search by `old_path` or `new_path`.
- Clicking on the `ID` takes the admin to the edit view. Clicking on `old_path` or `new_path` will take you to the URL (`old_path` should always redirect to `new_path` once this feature is complete - a quick way for an admin to test).
- Clicking on the `Date` column header will sort the rows by `Date`.
<img width="1179" alt="index_show_view" src="https://user-images.githubusercontent.com/15987080/83277886-b57e1600-a1a0-11ea-855e-5ae386eb327d.png">

**The create view**
<img width="1179" alt="create_view" src="https://user-images.githubusercontent.com/15987080/83277866-ae570800-a1a0-11ea-92f9-9ce4a0638196.png">

**The edit view**
<img width="1180" alt="edit_view" src="https://user-images.githubusercontent.com/15987080/83277849-a9925400-a1a0-11ea-84d5-950c692242b7.png">

## Added tests?

## Added to documentation?
**I'll add documentation on the final PR for this feature**

![dancing_air_traffic_controller](https://media.giphy.com/media/26FPt1fG0ioYIR35C/giphy.gif)
